README.md
# terraform-shell-azure-devops-elasticpool

[![GitHub Super-Linter](https://github.com/tonyskidmore/terraform-shell-azure-devops-elasticpool/workflows/Lint%20Code%20Base/badge.svg)](https://github.com/marketplace/actions/super-linter)

Azure DevOps VM Scale Set Elasticpool Terraform module.

Due to the fact that creating an [Agent Pool - Azure virtual machine scale set][scale-agents] is currently [blocked][blocking-issue]
due to not being supported by the SDK used by the [Azure DevOps Terraform Provider][terraform-provider-azuredevops],
this module uses the [Terraform shell provider][shell-provider] as a workaround.

This module is used by the [terraform-azurerm-vmss-devops-agent](https://registry.terraform.io/modules/tonyskidmore/vmss-devops-agent/azurerm/latest)
to create the Azure DevOps Scale Set elasticpool side of a self-hosted Azure DevOps Scale Set agent pool.

## Requirements

* An Azure subscription.
  _Note:_ you can get started with a [Azure free account][azure-free]

* An [Azure DevOps][azdo] [Organization][azdo-org].
  _Note:_ you can sign up for free in the preceding link.

* An Azure Virtual Machine Scale Set.
  _Note_: a test can can be created by using the [terraform-azurerm-vmss example][tf-vmss-example]

* An [Azure][azdo-connect-azure] service connection to the subscription where your Azure Virtual Machine Scale Set is located.

* An [Azure DevOps][azdo-project] project.

* An Azure DevOps [Personal Access Token][azdo-pat](PAT) created with at least Agent Pools (Read & manage) and Service Connections (Read & query) permissions.

* A Linux based system is required to execute this Terraform module, with the following commands installed:
  - cat
  - curl
  - sed
  - jq

The PAT needs be passed to the Terraform configuration by any standard mechanism, for example:

````bash

 export AZURE_DEVOPS_EXT_PAT="ckusfcc8ope2soot1yuovmdvlgtfgj9nio2orfwyvv5jsgcnwwga"
export TF_VAR_ado_ext_pat="$AZURE_DEVOPS_EXT_PAT"

````

_Note:_ The PAT is used for the initial creation of the agent pool and for subsequent Terraform operations.  Therefore, it would be advisable to create/use a service account for this rather than a standard user account.

<!-- BEGIN_TF_DOCS -->



## Basic example

```hcl

provider "azurerm" {
  features {}
}

provider "shell" {
  sensitive_environment = {
    AZURE_DEVOPS_EXT_PAT = var.ado_ext_pat
  }
}

data "azurerm_virtual_machine_scale_set" "ado_pool" {
  name                = var.vmss_name
  resource_group_name = var.vmss_resource_group_name
}

module "azure-devops-elasticpool" {
  source                 = "tonyskidmore/azure-devops-elasticpool/shell"
  version                = "0.4.4"
  ado_org                = var.ado_org
  ado_project            = var.ado_project
  ado_service_connection = var.ado_service_connection
  ado_pool_name          = var.ado_pool_name
  ado_pool_desired_idle  = var.ado_pool_desired_idle
  ado_vmss_id            = data.azurerm_virtual_machine_scale_set.ado_pool.id
}

```
## Resources

| Name | Type |
|------|------|
| [shell_script.ado_vmss_pool](https://registry.terraform.io/providers/scottwinkler/shell/latest/docs/resources/script) | resource |

## Inputs

| Name | Description | Type | Default | Required |
|------|-------------|------|---------|:--------:|
| <a name="input_ado_dirty"></a> [ado\_dirty](#input\_ado\_dirty) | Azure DevOps pool settings are dirty | `bool` | `false` | no |
| <a name="input_ado_org"></a> [ado\_org](#input\_ado\_org) | Azure DevOps Organization name | `string` | n/a | yes |
| <a name="input_ado_pool_auth_all_pipelines"></a> [ado\_pool\_auth\_all\_pipelines](#input\_ado\_pool\_auth\_all\_pipelines) | Setting to determine if all pipelines are authorized to use this TaskAgentPool by default (at create only) | `string` | `"True"` | no |
| <a name="input_ado_pool_auto_provision_projects"></a> [ado\_pool\_auto\_provision\_projects](#input\_ado\_pool\_auto\_provision\_projects) | Setting to automatically provision TaskAgentQueues in every project for the new pool (at create only) | `string` | `"True"` | no |
| <a name="input_ado_pool_desired_idle"></a> [ado\_pool\_desired\_idle](#input\_ado\_pool\_desired\_idle) | Number of machines to have ready waiting for jobs | `number` | `0` | no |
| <a name="input_ado_pool_desired_size"></a> [ado\_pool\_desired\_size](#input\_ado\_pool\_desired\_size) | The desired size of the pool | `number` | `0` | no |
| <a name="input_ado_pool_max_capacity"></a> [ado\_pool\_max\_capacity](#input\_ado\_pool\_max\_capacity) | Maximum number of machines that will exist in the elastic pool | `number` | `2` | no |
| <a name="input_ado_pool_max_saved_node_count"></a> [ado\_pool\_max\_saved\_node\_count](#input\_ado\_pool\_max\_saved\_node\_count) | Keep machines in the pool on failure for investigation | `number` | `0` | no |
| <a name="input_ado_pool_name"></a> [ado\_pool\_name](#input\_ado\_pool\_name) | Azure DevOps agent pool name | `string` | `"azdo-vmss-pool-001"` | no |
| <a name="input_ado_pool_os_type"></a> [ado\_pool\_os\_type](#input\_ado\_pool\_os\_type) | Operating system type of the nodes in the pool | `string` | `"linux"` | no |
| <a name="input_ado_pool_recycle_after_use"></a> [ado\_pool\_recycle\_after\_use](#input\_ado\_pool\_recycle\_after\_use) | Discard machines after each job completes | `bool` | `false` | no |
| <a name="input_ado_pool_sizing_attempts"></a> [ado\_pool\_sizing\_attempts](#input\_ado\_pool\_sizing\_attempts) | The number of sizing attempts executed while trying to achieve a desired size | `number` | `0` | no |
| <a name="input_ado_pool_ttl_mins"></a> [ado\_pool\_ttl\_mins](#input\_ado\_pool\_ttl\_mins) | The minimum time in minutes to keep idle agents alive | `number` | `30` | no |
| <a name="input_ado_project"></a> [ado\_project](#input\_ado\_project) | Azure DevOps project name where service connection exists and optionally where pool will only be created | `string` | n/a | yes |
| <a name="input_ado_project_only"></a> [ado\_project\_only](#input\_ado\_project\_only) | Only create the agent pool in the Azure DevOps pool specified? (at create only) | `string` | `"False"` | no |
| <a name="input_ado_service_connection"></a> [ado\_service\_connection](#input\_ado\_service\_connection) | Azure DevOps azure service connection name | `string` | n/a | yes |
| <a name="input_ado_vmss_id"></a> [ado\_vmss\_id](#input\_ado\_vmss\_id) | Azure Virtual Machine Scale Set Resource ID if not created by the module | `string` | `""` | no |
| <a name="input_http_connect_timeout"></a> [http\_connect\_timeout](#input\_http\_connect\_timeout) | The maximum time in seconds before timing out a connection to the Azure DevOps REST API | `number` | `20` | no |
| <a name="input_http_max_time"></a> [http\_max\_time](#input\_http\_max\_time) | The maximum amount of time in seconds for an Azure DevOps REST API operation to complete | `number` | `120` | no |
| <a name="input_http_retries"></a> [http\_retries](#input\_http\_retries) | The number of retries make to the Azure DevOps REST API | `number` | `10` | no |
| <a name="input_http_retries_max_time"></a> [http\_retries\_max\_time](#input\_http\_retries\_max\_time) | The maximum time in seconds for the retry period for a connection to the Azure DevOps REST API | `number` | `120` | no |
| <a name="input_http_retry_delay"></a> [http\_retry\_delay](#input\_http\_retry\_delay) | The maximum time in seconds to delay before retrying a connection to the Azure DevOps REST API | `number` | `3` | no |

## Outputs

| Name | Description |
|------|-------------|
| <a name="output_ado_vmss_pool_output"></a> [ado\_vmss\_pool\_output](#output\_ado\_vmss\_pool\_output) | Azure DevOps VMSS Agent Pool output |

## Providers

| Name | Version |
|------|---------|
| <a name="provider_shell"></a> [shell](#provider\_shell) | 1.7.10 |


<!-- END_TF_DOCS -->

## Troubleshooting

If an error occurs when running the module and the error is not immediately obvious then enabling a trace log prior to running the module can help to troubleshoot:

````bash

export TF_LOG=TRACE
export TF_LOG_PATH="./trace.log"

````

After enabling the trace log and recreating the issue look in the `trace.log` file and search for `[DEBUG] Starting execution...`.  Hopefully the additional output will provide additional information to help determine the underlying cause.

[scale-agents]: https://learn.microsoft.com/en-us/azure/devops/pipelines/agents/scale-set-agents
[shell-provider]: https://registry.terraform.io/providers/scottwinkler/shell/1.7.10
[blocking-issue]: https://github.com/microsoft/terraform-provider-azuredevops/issues/204
[terraform-provider-azuredevops]: https://github.com/microsoft/terraform-provider-azuredevops
[azdo-pat]: https://learn.microsoft.com/en-us/azure/devops/organizations/accounts/use-personal-access-tokens-to-authenticate
[azure-free]: https://azure.microsoft.com/en-gb/free
[azdo]: https://azure.microsoft.com/en-gb/products/devops
[azdo-org]: https://learn.microsoft.com/en-us/azure/devops/organizations/accounts/create-organization
[azdo-project]: https://learn.microsoft.com/en-us/azure/devops/organizations/projects/create-project
[azdo-connect-azure]: https://learn.microsoft.com/en-us/azure/devops/pipelines/library/connect-to-azure
[tf-vmss-example]: https://github.com/tonyskidmore/terraform-azurerm-vmss/tree/main/examples/admin_password


CHANGELOG.md
# CHANGELOG

[keepachangelog](https://keepachangelog.com/en/1.0.0/)

## [0.4.4]
* minor documentation updates

## [0.4.3]
* Checkov updates

## [0.4.2]
* Setting dirty as a variable

## [0.4.1]
* Fixing missing line continuation on curl

## [0.4.0]
* Adding curl control parameters

## [0.3.0]
* Fixing [#8](https://github.com/tonyskidmore/terraform-shell-azure-devops-elasticpool/issues/8)

## [0.2.0]
* Moved provider block
* Added unit tests to pre-commit
* Updated example

## [0.1.0]
Initial version


test/conftest.py
""" pytest configuration module. """

import os

import pytest
import tftest

BASEDIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'fixtures')

# setup parametr defaulst to True
# cleanup_on_exit=False

# @pytest.fixture(scope='session')
# def _plan_apply_runner():
#     "Returns a function to run Terraform plan and apply on a fixture."

#     # with tempfile.NamedTemporaryFile() as fp:
#     #   fp.close()

#     def run_plan(fixture_path, targets=None, refresh=True,
#                  tf_var_file=None, tf_vars=None, **kw):
#         """ Runs Terraform plan and apply and returns parsed output. """
#         tf_test = tftest.TerraformTest(fixture_path, BASEDIR,
#                                        os.environ.get('TERRAFORM',
#                                                       'terraform'))
#         tf_test.setup()
#         return tf_test.plan(output=False, refresh=refresh, tf_vars=tf_vars,
#                             targets=targets, tf_var_file=tf_var_file)
#     return run_plan


# @pytest.fixture(scope='session')
# def plan_runner(_plan_runner):
#     """ Returns a function to run Terraform plan on a module fixture. """

#     def run_plan(fixture_path, targets=None, tf_var_file=None, **tf_vars):
#         """ Runs Terraform plan and returns plan and module resources. """
#         plan = _plan_runner(fixture_path, targets=targets,
#                             tf_var_file=tf_var_file, **tf_vars)

#         root_module = plan.root_module['child_modules'][0]
#         return plan, root_module['resources']

#     return run_plan

@pytest.fixture(scope='session')
def _plan_runner():
    "Returns a function to run Terraform plan on a fixture."

    def run_plan(fixture_path, targets=None, refresh=True,
                 tf_var_file=None, **tf_vars):
        """ Runs Terraform plan and returns parsed output. """
        tf_test = tftest.TerraformTest(fixture_path, BASEDIR,
                                       os.environ.get('TERRAFORM',
                                                      'terraform'))
        tf_test.setup(cleanup_on_exit=False)
        return tf_test.plan(output=True, refresh=refresh, tf_vars=tf_vars,
                            targets=targets, tf_var_file=tf_var_file)
    return run_plan


@pytest.fixture(scope='session')
def plan_runner(_plan_runner):
    """ Returns a function to run Terraform plan on a module fixture. """

    def run_plan(fixture_path, targets=None, tf_var_file=None, **tf_vars):
        """ Runs Terraform plan and returns plan and module resources. """
        plan = _plan_runner(fixture_path, targets=targets,
                            tf_var_file=tf_var_file, **tf_vars)

        root_module = plan.root_module['child_modules'][0]
        return plan, root_module['resources']

    return run_plan


@pytest.fixture(scope='session')
def plan_apply_runner():
    """ Returns a function to run Terraform plan and apply
    on a module fixture. """

    def run_play_apply(fixture_path,
                       targets=None,
                       refresh=True,
                       tf_var_file=None, **tf_vars):
        """ Runs Terraform plan and applt and returns parsed output. """

        tf_test = tftest.TerraformTest(fixture_path, BASEDIR,
                                       os.environ.get('TERRAFORM',
                                                      'terraform'))
        tf_test.setup(cleanup_on_exit=False)
        plan = tf_test.plan(output=True, refresh=refresh, tf_vars=tf_vars,
                            targets=targets, tf_var_file=tf_var_file)
        apply = tf_test.apply(tf_vars=tf_vars, tf_var_file=tf_var_file)
        output = tf_test.output(json_format=True)

        return plan, apply, output

    return run_play_apply


@pytest.fixture(scope='session')
def e2e_plan_runner(_plan_runner):
    """ Returns a function to run Terraform plan on an end-to-end fixture. """

    def run_plan(fixture_path, targets=None, refresh=True,
                 tf_var_file=None, **tf_vars):
        """ Runs Terraform plan on an end-to-end module using defaults. """
        plan = _plan_runner(fixture_path, targets=targets, refresh=refresh,
                            tf_var_file=tf_var_file, **tf_vars)
        # skip the fixture
        root_module = plan.root_module['child_modules'][0]
        modules = dict((mod['address'], mod['resources'])
                       for mod in root_module['child_modules'])
        resources = [r for m in modules.values() for r in m]
        return plan, modules, resources

    return run_plan


@pytest.fixture(scope='session')
def example_plan_runner(_plan_runner):
    """ Returns a function to run Terraform plan on documentation examples. """

    def run_plan(fixture_path):
        """ Runs Terraform plan and returns count of modules and resources. """
        plan = _plan_runner(fixture_path)
        # the fixture is the example we are testing
        return (
            len(plan.modules),
            sum(len(m.resources) for m in plan.modules.values()))

    return run_plan


@pytest.fixture(scope='session')
def apply_runner():
    """ Returns a function to run Terraform apply on a fixture. """

    def run_apply(fixture_path, tf_var_file=None, **tf_vars):
        """ Runs Terraform apply and returns parsed output. """

        tf_test = tftest.TerraformTest(fixture_path, BASEDIR,
                                       os.environ.get('TERRAFORM',
                                                      'terraform'))
        tf_test.setup(cleanup_on_exit=False)
        apply = tf_test.apply(tf_vars=tf_vars, tf_var_file=tf_var_file)
        output = tf_test.output(json_format=True)
        return apply, output

    return run_apply


@pytest.fixture(scope='session')
def destroy_runner():
    """ Returns a function to run Terraform destroy on a fixture. """

    def run_destroy(fixture_path, tf_var_file=None, **tf_vars):
        """ Runs Terraform destroy and returns parsed output. """

        tf_test = tftest.TerraformTest(fixture_path, BASEDIR,
                                       os.environ.get('TERRAFORM',
                                                      'terraform'))
        tf_test.setup(cleanup_on_exit=True)
        output = tf_test.destroy(tf_vars=tf_vars, tf_var_file=tf_var_file,
                                 **{"auto_approve": True})
        return output

    return run_destroy


test/check_ado_pool.py
""" Helper script for calling get_ado_pool """

import os

from ado.get_ado_pool import get_ado_pool

org = os.environ['TF_VAR_ado_org']
pool_id = os.environ['TF_VAR_pool_id']
pat = os.environ['TF_VAR_ado_ext_pat']

resp = get_ado_pool(org, pool_id, pat)
print(resp)


test/ado/get_ado_pool.py
""" Gets settings of an Azure DevOps ElasticPool. """

# export TF_VAR_ado_org="https://dev.azure.com/tonyskidmore"
# export TF_VAR_pool_id="47"
# export TF_VAR_ado_ext_pat="$AZURE_DEVOPS_EXT_PAT"

import os

import requests
from requests.auth import HTTPBasicAuth


def get_ado_pool(org, pool_id, pat):
    """ Get Azure DevOps pool settings. """

    api_url = (
        f"{org}/_apis/distributedtask/elasticpools/"
        f"{pool_id}?api-version=7.1-preview.1"
    )

    headers = {'Accept': 'application/json'}

    try:
        response = requests.get(api_url, headers=headers,
                                auth=HTTPBasicAuth('', pat),
                                timeout=10)
        response.raise_for_status()
    except requests.exceptions.HTTPError as errh:
        print("Http Error:", errh)
    except requests.exceptions.ConnectionError as errc:
        print("Error Connecting:", errc)
    except requests.exceptions.Timeout as errt:
        print("Timeout Error:", errt)
    except requests.exceptions.RequestException as err:
        print("Requests Exception Error", err)

    return response.status_code, response.json()


def main():
    """ main function. """

    try:
        org = os.environ['TF_VAR_ado_org']
        pool_id = os.environ['TF_VAR_pool_id']  # output.poolId
        pat = os.environ['TF_VAR_ado_ext_pat']
    except KeyError as err:
        print("An error occurred:", err)
        raise err

    _, pool_settings = get_ado_pool(org, pool_id, pat)
    print(pool_settings)


if __name__ == "__main__":
    main()


scripts/read.sh
#!/bin/sh

set -eu

# functions

curlwithcode() {

  status=0
  output="output.out"
  out=""

  # Run curl in a separate command, capturing output of -w "%{http_code}" into statuscode
  # and sending the content to a file with -o >(cat >/tmp/curl_body)
  http_code=$(curl \
                --silent \
                --show-error \
                --write-out '%{http_code}' \
                --header "Content-Type: application/json" \
                --request "$1" \
                --user ":$AZURE_DEVOPS_EXT_PAT" \
                --data @params.json \
                --output "$output" \
                "$2"
  ) || status="$?"

  out=$(cat "$output")
  rm -f "$output"

}

checkout() {

  if [ "$http_code" != "200" ]
  then
    if [ "$(echo "$out" | jq empty > /dev/null 2>&1; echo $?)" = "0" ]
    then
      echo "Parsed JSON successfully and got something other than false/null"
      message="$(echo "$out" | jq -r '.message')"
      printf "Error: %s\n" "$message" >&2
      exit 1
    else
      printf "Failed to parse JSON, or got false/null.\n" >&2
      if echo "$out" | grep -q "_signin"
      then
        printf "Azure DevOps PAT token is not correct\n" >&2
        exit 1
      elif echo "$out" | grep -q "The resource cannot be found."
      then
        printf "The resource cannot be found.\n" >&2
        exit 1
      else
        printf "%s\n" "$out"
        exit 1
      fi
    fi
  elif [ "$status" != "0" ]
  then
    printf "status: %s\n" "$status"  >&2
    printf "%s\n" "$out"
    exit 1
  fi

}

# end functions

std_in=$(cat)
printf "std_in: %s\n" "$std_in"
poolId=$(echo "$std_in" | jq -r '.poolId')
printf "poolId: %s\n" "$poolId"

# https://learn.microsoft.com/en-us/rest/api/azure/devops/distributedtask/elasticpools/get?view=azure-devops-rest-7.1
poolUrl="$ADO_ORG/_apis/distributedtask/elasticpools/${poolId}?api-version=7.1-preview.1"

curlwithcode "GET" "$poolUrl"
checkout

printf "%s" "$out" | jq -r 'del(.offlineSince, .state)'


scripts/ado_elastic_pool.sh
#!/bin/bash

main () {
  check_env_vars
  prereqs
  printf "mode: %s\n" "$mode"
  "${mode}_func"
}

# different source paths when running bats tests
if [[ -z "$BATS_TEST_FILENAME" ]]
then
  script_dir="$(dirname "$(realpath "$0")")"
  # shellcheck source=$script_dir/functions.sh
  # shellcheck disable=SC1091
  source "$script_dir/functions.sh"
else
  # shellcheck source=./scripts/functions.sh
  # shellcheck disable=SC1091
  source ./scripts/functions.sh
fi

# similar method in bash to the Python:
# if __name__ == "__main__":
#     main()
# to allow the distinction of being called from the command line or via source
if [[ "${BASH_SOURCE[0]}" == "${0}" ]]
then
  mode="$1"
  case $mode in
    create|read|update|delete)
      main ;;
    *)
      echo "Only accepts create, read, update or delete" ; exit 1;;
  esac
fi


scripts/functions.sh
#!/bin/bash

create_func () {

  # Crud - Create operation

  get_projects
  get_endpoint
  build_pool_url

  # Create the elasticpool
  rest_api_call "POST" "$poolUrl"

  # get the poolId of the new elasticpool
  pool_id=$(echo "$out" | jq -r .elasticPool.poolId)
  # switch to read mode and store details in state
  mode="read"
  get_elasticpool_by_id

}


read_func() {

  # cRud - Read operation

  # get current state
  input_state
  # update state
  get_elasticpool_by_id

}



update_func() {

  # crUd - Update operation

  # get current status
  input_state

  # build url and make call for update rest api method
  build_pool_url
  rest_api_call "PATCH" "$poolUrl"
  # update state
  get_elasticpool_by_id

}


delete_func() {

  # cruD - Delete operation

  # get current state
  input_state

  # build url and make call for delete rest api method
  build_pool_url
  rest_api_call "DELETE" "$poolUrl"

}


env_not_set() {
  raise "$1 must be set before executing this script"
  ((failed_count++))
}


check_env_vars() {

  failed_count=0
  [[ -z "$ADO_ORG" ]] && env_not_set "ADO_ORG"
  [[ -z "$ADO_POOL_AUTH_ALL_PIPELINES" ]] && env_not_set "ADO_POOL_AUTH_ALL_PIPELINES"
  [[ -z "$ADO_POOL_AUTO_PROVISION" ]] && env_not_set "ADO_POOL_AUTO_PROVISION"
  [[ -z "$ADO_POOL_DESIRED_IDLE" ]] && env_not_set "ADO_POOL_DESIRED_IDLE"
  [[ -z "$ADO_POOL_DESIRED_SIZE" ]] && env_not_set "ADO_POOL_DESIRED_SIZE"
  [[ -z "$ADO_POOL_MAX_CAPACITY" ]] && env_not_set "ADO_POOL_MAX_CAPACITY"
  [[ -z "$ADO_POOL_MAX_SAVED_NODE_COUNT" ]] && env_not_set "ADO_POOL_MAX_SAVED_NODE_COUNT"
  [[ -z "$ADO_POOL_NAME" ]] && env_not_set "ADO_POOL_NAME"
  [[ -z "$ADO_POOL_OS_TYPE" ]] && env_not_set "ADO_POOL_OS_TYPE"
  [[ -z "$ADO_POOL_RECYCLE_AFTER_USE" ]] && env_not_set "ADO_POOL_RECYCLE_AFTER_USE"
  [[ -z "$ADO_POOL_SIZING_ATTEMPTS" ]] && env_not_set "ADO_POOL_SIZING_ATTEMPTS"
  [[ -z "$ADO_POOL_TTL_MINS" ]] && env_not_set "ADO_POOL_TTL_MINS"
  [[ -z "$ADO_PROJECT" ]] && env_not_set "ADO_PROJECT"
  [[ -z "$ADO_PROJECT_ONLY" ]] && env_not_set "ADO_PROJECT_ONLY"
  [[ -z "$ADO_SERVICE_CONNECTION" ]] && env_not_set "ADO_SERVICE_CONNECTION"
  [[ -z "$AZ_VMSS_ID" ]] && env_not_set "AZ_VMSS_ID"
  [[ -z "$AZURE_DEVOPS_EXT_PAT" ]] && env_not_set "AZURE_DEVOPS_EXT_PAT"
  [[ $failed_count -ne 0 ]] && exit 1

}


input_state() {
  # read in state from stdin
  std_in=$(cat)
  printf "std_in: %s\n" "$std_in"
  pool_id=$(echo "$std_in" | jq -r '.poolId')
  endpoint_id=$(echo "$std_in" | jq -r '.serviceEndpointId')
  project_id=$(echo "$std_in" | jq -r '.serviceEndpointScope')
}


# output_state
output_state() {
  # save state from stdout
  printf "%s" "$out" | jq -r 'del(.offlineSince, .state)'
}


get_projects() {

  # Get ADO projects to allow obtaining required project ID
  # https://learn.microsoft.com/en-us/rest/api/azure/devops/core/projects/get?view=azure-devops-rest-6.0
  projectUrl="$ADO_ORG/_apis/projects?api-version=6.0"
  rest_api_call "GET" "$projectUrl"
  project=$(printf "%s" "$out" | jq -r --arg name "$ADO_PROJECT" '.value[] | select (.name==$name)')
  project_id=$(echo "$project" | jq -r '.id')
  if [[ -n "$project_id" ]]
  then
    printf "project_id: %s\n" "$project_id"
  else
    printf "Failed to obtain project_id for project: %s\n" "$ADO_PROJECT" >&2
    exit 1
  fi

}


get_endpoint() {

  # Get endpoint ID of the specified service connection in the target project
  # https://learn.microsoft.com/en-us/rest/api/azure/devops/serviceendpoint/endpoints/get-service-endpoints-by-names?view=azure-devops-rest-6.0&tabs=HTTP
  endpointUrl="$ADO_ORG/$ADO_PROJECT/_apis/serviceendpoint/endpoints?endpointNames=$ADO_SERVICE_CONNECTION&api-version=6.0-preview.4"
  rest_api_call "GET" "$endpointUrl"
  endpoint_id=$(echo "$out" | jq -r '.value[].id')
  if [[ -n "$endpoint_id" ]]
  then
    printf "endpoint_id: %s\n" "$endpoint_id"
  else
    printf "Failed to obtain endpoint_id for service connection: %s\n" "$ADO_SERVICE_CONNECTION" >&2
    exit 1
  fi

}


build_pool_url() {

  # https://learn.microsoft.com/en-us/rest/api/azure/devops/distributedtask/elasticpools?view=azure-devops-rest-7.1
  # mode is global varaiable defined in parent script
  # shellcheck disable=SC2154
  if [[ "$mode" == "read" || "$mode" == "update" ]]
  then
    poolUrl="${ADO_ORG}/_apis/distributedtask/elasticpools/$pool_id?api-version=7.1-preview.1"
  elif [[ "$mode" == "delete" ]]
  then
    # https://learn.microsoft.com/en-us/rest/api/azure/devops/distributedtask/agents/delete?view=azure-devops-rest-7.1
    poolUrl="${ADO_ORG}/_apis/distributedtask/pools/$pool_id?api-version=7.1-preview.1"
  elif [[ "$mode" == "create" ]]
  then
    # Only specify the optional project ID if the pool is only required in the specified project, determined by if ADO_PROJECT_ONLY is True
    # https://learn.microsoft.com/en-us/rest/api/azure/devops/distributedtask/elasticpools/create?view=azure-devops-rest-7.1
    url_prefix="${ADO_ORG}/_apis/distributedtask/elasticpools?poolName=${ADO_POOL_NAME}&authorizeAllPipelines=${ADO_POOL_AUTH_ALL_PIPELINES}&autoProvisionProjectPools=${ADO_POOL_AUTO_PROVISION}"
    url_suffix="&api-version=7.1-preview.1"
    if [[ "$ADO_PROJECT_ONLY" == "True" ]]
    then
      url_suffix="&projectId=${project_id}${url_suffix}"
    fi
    poolUrl="${url_prefix}${url_suffix}"
  else
    raise "Failed to build pool URL"
    exit 1
  fi

}


get_elasticpool_by_id() {

  build_pool_url
  rest_api_call "GET" "$poolUrl"
  output_state

}


create_post_data()
{
  cat <<EOF
{
  "agentInteractiveUI": false,
  "azureId": "$AZ_VMSS_ID",
  "desiredIdle": $ADO_POOL_DESIRED_IDLE,
  "desiredSize": $ADO_POOL_DESIRED_SIZE,
  "maxCapacity": $ADO_POOL_MAX_CAPACITY,
  "maxSavedNodeCount": $ADO_POOL_MAX_SAVED_NODE_COUNT,
  "offlineSince": "",
  "osType": "$ADO_POOL_OS_TYPE",
  "recycleAfterEachUse": $ADO_POOL_RECYCLE_AFTER_USE,
  "serviceEndpointId": "$endpoint_id",
  "serviceEndpointScope": "$project_id",
  "sizingAttempts": $ADO_POOL_SIZING_ATTEMPTS,
  "state": "online",
  "timeToLiveMinutes": $ADO_POOL_TTL_MINS
}
EOF
}


update_post_data()
{
  cat <<EOF
{
  "agentInteractiveUI": false,
  "azureId": "$AZ_VMSS_ID",
  "desiredIdle": $ADO_POOL_DESIRED_IDLE,
  "maxCapacity": $ADO_POOL_MAX_CAPACITY,
  "maxSavedNodeCount": $ADO_POOL_MAX_SAVED_NODE_COUNT,
  "osType": "$ADO_POOL_OS_TYPE",
  "recycleAfterEachUse": $ADO_POOL_RECYCLE_AFTER_USE,
  "serviceEndpointId": "$endpoint_id",
  "serviceEndpointScope": "$project_id",
  "timeToLiveMinutes": $ADO_POOL_TTL_MINS
}
EOF
}


raise() {
  printf "%s\n" "$1" >&2
}


check_command () {
  # Determine if command is installed
  command -v "${1}" &>/dev/null
}


check_prereqs() {

  prereqs=("$@")

  for cmd in "${prereqs[@]}"
  do
    if ! check_command "$cmd"
    then
      raise "Module prerequisite not installed: $cmd"
      exit 6
    fi
  done
}


prereqs() {
  cmds=("jq" "curl" "cat" "sed")
  check_prereqs "${cmds[@]}"
}


build_params() {

  local method="$1"
  local url="$2"

  params=(
          "--silent" \
          "--show-error" \
          "--retry" "$HTTP_RETRIES" \
          "--retry-delay" "$HTTP_RETRY_DELAY" \
          "--retry-max-time" "$HTTP_RETRIES_MAX_TIME" \
          "--max-time" "$HTTP_MAX_TIME" \
          "--connect-timeout" "$HTTP_CONNECT_TIMEOUT" \
          "--write-out" "\n%{http_code}" \
          "--header" "Content-Type: application/json" \
          "--request" "$method"
  )

  if [[ "$method" == "POST" || "$method" == "PATCH" ]]
  then
    # generate curl --data if method if POST(create) or PATCH(update)
    # mode is global varaiable defined in parent script
    # shellcheck disable=SC2154
    data_func="${mode}_post_data"
    data="$($data_func)"
    printf "data: %s\n" "$data"
    params+=("--data" "$data")
  fi

  params+=("--user" ":$AZURE_DEVOPS_EXT_PAT" "$url")

}


rest_api_call() {

  exit_code=0

  if [ "$#" -ne 2 ]
  then
      printf "Expected 2 function arguments, got %s\n" "$#"
      exit 1
  fi

  method="$1"
  url="$2"


  if [[ "$method" == "GET" || "$method" == "POST" || "$method" == "PATCH" || "$method" == "DELETE" ]]
  then
    printf "method: %s\n" "$method"
  else
    printf "Expected method to be one of: GET,POST,PATCH.DELETE got %s\n" "$method"
    exit 1
  fi

  if [[ $url =~ ^https:\/\/.+\/_apis.+$ ]]
  then
    printf "url: %s\n" "$url"
  else
    printf "Invalid or missing URL: %s\n" "$url"
    exit 1
  fi

  build_params "$method" "$url"

  printf "curl %s\n" "${params[*]}"
  res=$(curl "${params[@]}")
  exit_code=$?

  # https://unix.stackexchange.com/questions/572424/retrieve-both-http-status-code-and-content-from-curl-in-a-shell-script
  http_code=$(tail -n1 <<< "$res") # get the last line
  out=$(sed '$ d' <<< "$res") # get all but the last line which contains the status code

  printf "http_code: %s\n" "$http_code"
  printf "out: %s\n" "$out"

  checkout

}


checkout() {

  if [[ "$exit_code" != "0" ]]
  then
    raise "Operation failed. Mode: $mode, Method: $method, exit_code: $exit_code, HTTP code: $http_code"
    printf "%s\n" "$out"
    exit 1
  else
    echo "out"
    if [[ "$mode" != "delete" && "$http_code" == "200" ]] || [[ "$mode" == "delete" && "$http_code" == "204" ]]
    then
      regex="_apis\/([a-z]+)"
      [[ $url =~ $regex ]] && service="${BASH_REMATCH[1]}"
      printf "Operation successful. Service: %s, Mode: %s, Method: %s, exit_code: %s, HTTP code: %s\n" "$service" "$mode" "$method" "$exit_code" "$http_code"
    else
      if [[ -n "$out" ]] && [[ "$(echo "$out" | jq empty > /dev/null 2>&1; echo $?)" = "0" ]]
      then
        printf "Parsed JSON successfully and got something other than false/null\n"
        message="$(echo "$out" | jq -r '.message')"
        raise "Error: $message"
        exit 2
      else
        printf "Failed to parse JSON, or got false/null\n"
        if echo "$out" | grep -q "_signin"
        then
          raise "Azure DevOps PAT token is not correct"
          exit 4
        elif echo "$out" | grep -q "The resource cannot be found"
        then
          raise "The resource cannot be found"
          exit 5
        elif [[ "$http_code" == "401" ]]
        then
          raise "401 Unauthorized, probable PAT permissions issue"
          exit 7
        else
        raise "Unknown error"
          printf "%s\n" "$out"
          exit 3
        fi
      fi
    fi
  fi

}
